packages/credential-provider-imds/src/fromInstanceMetadata.ts

```diff
@@ -154,23 +200,113 @@ const getMetadataToken = async (options: RequestOptions) =>
+/**
+ * @internal
+ * Checks if IMDS credential fetching is disabled through configuration
+ */
+export const throwIfImdsTurnedOff = async (profile?: string, logger?: any): Promise<void> => {
+  // Load configuration in priority order
+  const disableImds = await loadConfig(
+    {
+      // Check environment variable
+      environmentVariableSelector: (env) => {
+        const envValue = env[ENV_IMDS_DISABLED];
+        if (envValue === undefined) {
+          return undefined;
+        }
+        return envValue === "true";
+      },
+      // Check config file
+      configFileSelector: (profile) => {
+        const profileValue = profile[CONFIG_IMDS_DISABLED];
+        if (profileValue === undefined) {
+          return undefined;
+        }
+        return profileValue === "true";
+      },
+      default: false,
+    },
+    { profile }
+  )();
+
+  // If IMDS is disabled, throw error
+  if (disableImds) {
+    throw new CredentialsProviderError("IMDS credential fetching is disabled", { logger });
+  }
+};
+
+/**
+ * @internal
+ * Gets configured profile name from various sources
+ */
+export const getConfiguredProfileName = async (init: RemoteProviderInit, profile?: string): Promise<string | null> => {
+  // Load configuration in priority order
+  const profileName = await loadConfig(
+    {
+      // Check environment variable
+      environmentVariableSelector: (env) => env[ENV_PROFILE_NAME],
+      // Check config file
+      configFileSelector: (profile) => profile[CONFIG_PROFILE_NAME],
+      default: null,
+    },
+    { profile }
+  )();
+
+  // Check runtime config (highest priority)
+  const name = init.ec2InstanceProfileName ?? profileName;
+
+  // Validate if name is provided but empty
+  if (typeof name === "string" && name.trim() === "") {
+    throw new CredentialsProviderError("EC2 instance profile name cannot be empty");
+  }
+
+  return name;
+};
 
+/**
+ * @internal
+ * Gets credentials from profile
+ */
 const getCredentialsFromProfile = async (profile: string, options: RequestOptions, init: RemoteProviderInit) => {
-  const credentialsResponse = JSON.parse(
-    (
-      await httpRequest({
-        ...options,
-        path: IMDS_PATH + profile,
-      })
-    ).toString()
-  );
+  // Try extended API first
+  try {
+    return await getCredentialsFromPath(IMDS_EXTENDED_PATH + profile, options);
+  } catch (error) {
+    // If extended API returns 404, fall back to legacy API
+    if (error.statusCode === 404) {
+      try {
+        return await getCredentialsFromPath(IMDS_LEGACY_PATH + profile, options);
+      } catch (legacyError) {
+        if (legacyError.statusCode === 404 && init.ec2InstanceProfileName === undefined) {
+          // If legacy API also returns 404 and we're using a cached profile name,
+          // the profile might have changed - clear cache and retry
+          const newProfileName = await getImdsProfile(options, init.maxRetries ?? 3, init, profile, true);
```

Comment:
how does `init.ec2InstanceProfileName === undefined` indicate that we're using a cached profile name?